design_docs/parallel_tasks.md

```diff
@@ -0,0 +1,396 @@
+                taskName, logs[taskName].name)
+        logs[taskName].close()
+        # remove the process from the process dictionary (no need to bother)
+        processes.pop(taskName)
+
+        if len(remainingTasks) > 0:
+            (process, log) = launch_tasks(remainingTasks[0:1], config,
+                                          configFiles)
+            # merge the new process and log into these dictionaries
+            processes.update(process)
+            logs.update(log)
+            remainingTasks = remainingTasks[1:]
+    # }}}
+
+
+def launch_tasks(taskNames, config, configFiles):  # {{{
+    """
+    Launch one or more tasks
+
+    Author: Xylar Asay-Davis
+    Last Modified: 03/08/2017
+    """
+    thisFile = os.path.realpath(__file__)
+
+    logsDirectory = build_config_full_path(config, 'output',
+                                           'logsSubdirectory')
+    make_directories(logsDirectory)
+
+    commandPrefix = config.getWithDefault('execute', 'commandPrefix',
+                                          default='')
+    if commandPrefix == '':
+        commandPrefix = []
+    else:
+        commandPrefix = commandPrefix.split(' ')
+
+    processes = {}
+    logs = {}
+    for taskName in taskNames:
+        args = commandPrefix + [thisFile, '--generate', taskName] + configFiles
+
+        logFileName = '{}/{}.log'.format(logsDirectory, taskName)
+
+        # write the command to the log file
+        logFile = open(logFileName, 'w')
+        logFile.write('Command: {}\n'.format(' '.join(args)))
+        # make sure the command gets written before the rest of the log
+        logFile.flush()
+        print 'Running {}'.format(taskName)
+        process = subprocess.Popen(args, stdout=logFile,
+                                   stderr=subprocess.STDOUT)
+        processes[taskName] = process
+        logs[taskName] = logFile
+
+    return (processes, logs)  # }}}
+
+
+def wait_for_task(processes):  # {{{
+    """
+    Wait for the next process to finish and check its status.  Returns both the
+    task name and the process that finished.
+
+    Author: Xylar Asay-Davis
+    Last Modified: 03/08/2017
+    """
+
+    # first, check if any process has already finished
+    for taskName, process in processes.iteritems():  # python 2.7!
+        if(not is_running(process)):
+            return (taskName, process)
+
+    # No process has already finished, so wait for the next one
+    (pid, status) = os.waitpid(-1, 0)
+    for taskName, process in processes.iteritems():
+        if pid == process.pid:
+            process.returncode = status
+            # since we used waitpid, this won't happen automatically
+            return (taskName, process)  # }}}
+
+
+def is_running(process):  # {{{
+    """
+    Returns whether a given process is currently running
+
+    Author: Xylar Asay-Davis
+    Last Modified: 03/08/2017
+    """
+
+    try:
+        os.kill(process.pid, 0)
```

Comment:
Is there a cleaner way to do this?  I thought `POpen` had this type of functionality already included, e.g., would `POpen.poll()` or `Popen.wait()` (https://docs.python.org/2/library/subprocess.html) work?